### ISSUE TITLE
Accumulo Store with Kerberos 

### ISSUE THREAD

**Issue by tdong6** [gchq/Gaffer#2348]
Hi,

I have an Accumulo cluster (version: 1.7.2-cdh5.5.0) with kerberos enabled and I would like to know how I can configure Gaffer to use my Accumulo cluster as the underlying store. 

I have read the [doc](https://gchq.github.io/gaffer-doc/stores/accumulo-store.html) on Accumulo store and looked at the `TheBasics.java` from `gaffer-doc` repo (master branch). I didn't find any instructions on whether specific settings are required for kerberos. 

I downloaded the iterators.jar and distributed it to all my tservers. Then I modified `TheBasics.java` with my Accumulo properties
```java
        final AccumuloProperties properties = new AccumuloProperties();
        properties.setStoreClass(AccumuloStore.class);
        properties.setInstance("<my_instance>");
        properties.setZookeepers("<my_zk1>:2181,<my_zk2>:2181,<my_zk3>:2181");
        properties.setUser("myuser@EXAMPLE.COM");
        properties.setPassword("<my_password>");
```
And I get the following Exception
```
Exception in thread "main" java.lang.IllegalArgumentException: Could not initialise the store with provided arguments.
	at uk.gov.gchq.gaffer.store.Store.createStore(Store.java:262)
	at uk.gov.gchq.gaffer.graph.Graph$Builder.updateStore(Graph.java:1066)
	at uk.gov.gchq.gaffer.graph.Graph$Builder.build(Graph.java:953)
	at uk.gov.gchq.gaffer.doc.user.walkthrough.TheBasics.run(TheBasics.java:121)
	at uk.gov.gchq.gaffer.doc.user.walkthrough.TheBasics.main(TheBasics.java:161)
Caused by: uk.gov.gchq.gaffer.store.StoreException: Failed to create accumulo connection
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.getConnector(TableUtils.java:219)
	at uk.gov.gchq.gaffer.accumulostore.AccumuloStore.getConnection(AccumuloStore.java:205)
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.ensureTableExists(TableUtils.java:76)
	at uk.gov.gchq.gaffer.accumulostore.AccumuloStore.initialise(AccumuloStore.java:160)
	at uk.gov.gchq.gaffer.store.Store.createStore(Store.java:260)
	... 4 more
Caused by: org.apache.accumulo.core.client.AccumuloException: java.lang.RuntimeException: Cannot use SASL if Hadoop security is not enabled
	at org.apache.accumulo.core.client.impl.ServerClient.execute(ServerClient.java:70)
	at org.apache.accumulo.core.client.impl.ConnectorImpl.<init>(ConnectorImpl.java:73)
	at org.apache.accumulo.core.client.ZooKeeperInstance.getConnector(ZooKeeperInstance.java:308)
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.getConnector(TableUtils.java:217)
	... 8 more
Caused by: java.lang.RuntimeException: Cannot use SASL if Hadoop security is not enabled
	at org.apache.accumulo.core.rpc.SaslConnectionParams.updatePrincipalFromUgi(SaslConnectionParams.java:194)
	at org.apache.accumulo.core.rpc.SaslConnectionParams.<init>(SaslConnectionParams.java:140)
	at org.apache.accumulo.core.client.impl.ClientContext$3.get(ClientContext.java:117)
	at org.apache.accumulo.core.client.impl.ClientContext$3.get(ClientContext.java:109)
	at com.google.common.base.Suppliers$ExpiringMemoizingSupplier.get(Suppliers.java:185)
	at org.apache.accumulo.core.client.impl.ClientContext.getSaslParams(ClientContext.java:182)
	at org.apache.accumulo.core.client.impl.ThriftTransportKey.<init>(ThriftTransportKey.java:44)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:144)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:124)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:119)
	at org.apache.accumulo.core.client.impl.ServerClient.executeRaw(ServerClient.java:100)
	at org.apache.accumulo.core.client.impl.ServerClient.execute(ServerClient.java:64)
	... 11 more

```
Any suggestions?

**Comment by d47853** [comment]
That appears to be failing after checking the authentication method which would imply that the either the hadoop.security.authentication is unset or is set to "simple" in the hadoop conf. This should be set to "kerberos" for your example.

```java
// SaslConnectionParams.java 
protected void updatePrincipalFromUgi() {
    // Ensure we're using Kerberos auth for Hadoop UGI
    if (!UserGroupInformation.isSecurityEnabled()) {
      throw new RuntimeException("Cannot use SASL if Hadoop security is not enabled");
    }
```

```java
// UserGroupInformation.java
 /**
   * Determine if UserGroupInformation is using Kerberos to determine
   * user identities or is relying on simple authentication
   * 
   * @return true if UGI is working in a secure environment
   */
  public static boolean isSecurityEnabled() {
    return !isAuthenticationMethodEnabled(AuthenticationMethod.SIMPLE);
  }
  
  @InterfaceAudience.Private
  @InterfaceStability.Evolving
  private static boolean isAuthenticationMethodEnabled(AuthenticationMethod method) {
    ensureInitialized();
    return (authenticationMethod == method);
  }
```

**Comment by tdong6** [comment]
@d47853 Thanks for your suggestions. 

I set the value of `hadoop.security.authentication` to be `kerberos` and got a new exception:
```
Exception in thread "main" java.lang.IllegalArgumentException: Could not initialise the store with provided arguments.
	at uk.gov.gchq.gaffer.store.Store.createStore(Store.java:262)
	at uk.gov.gchq.gaffer.graph.Graph$Builder.updateStore(Graph.java:1066)
	at uk.gov.gchq.gaffer.graph.Graph$Builder.build(Graph.java:953)
	at uk.gov.gchq.gaffer.doc.user.walkthrough.TheBasics.run(TheBasics.java:128)
	at uk.gov.gchq.gaffer.doc.user.walkthrough.TheBasics.main(TheBasics.java:168)
Caused by: uk.gov.gchq.gaffer.store.StoreException: Failed to create accumulo connection
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.getConnector(TableUtils.java:219)
	at uk.gov.gchq.gaffer.accumulostore.AccumuloStore.getConnection(AccumuloStore.java:205)
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.ensureTableExists(TableUtils.java:76)
	at uk.gov.gchq.gaffer.accumulostore.AccumuloStore.initialise(AccumuloStore.java:160)
	at uk.gov.gchq.gaffer.store.Store.createStore(Store.java:260)
	... 4 more
Caused by: org.apache.accumulo.core.client.AccumuloException: java.lang.IllegalArgumentException: Cannot determine SASL mechanism for token class: class org.apache.accumulo.core.client.security.tokens.PasswordToken
	at org.apache.accumulo.core.client.impl.ServerClient.execute(ServerClient.java:70)
	at org.apache.accumulo.core.client.impl.ConnectorImpl.<init>(ConnectorImpl.java:73)
	at org.apache.accumulo.core.client.ZooKeeperInstance.getConnector(ZooKeeperInstance.java:308)
	at uk.gov.gchq.gaffer.accumulostore.utils.TableUtils.getConnector(TableUtils.java:217)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Cannot determine SASL mechanism for token class: class org.apache.accumulo.core.client.security.tokens.PasswordToken
	at org.apache.accumulo.core.rpc.SaslConnectionParams.updateFromToken(SaslConnectionParams.java:155)
	at org.apache.accumulo.core.rpc.SaslConnectionParams.<init>(SaslConnectionParams.java:142)
	at org.apache.accumulo.core.client.impl.ClientContext$3.get(ClientContext.java:117)
	at org.apache.accumulo.core.client.impl.ClientContext$3.get(ClientContext.java:109)
	at com.google.common.base.Suppliers$ExpiringMemoizingSupplier.get(Suppliers.java:185)
	at org.apache.accumulo.core.client.impl.ClientContext.getSaslParams(ClientContext.java:182)
	at org.apache.accumulo.core.client.impl.ThriftTransportKey.<init>(ThriftTransportKey.java:44)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:144)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:124)
	at org.apache.accumulo.core.client.impl.ServerClient.getConnection(ServerClient.java:119)
	at org.apache.accumulo.core.client.impl.ServerClient.executeRaw(ServerClient.java:100)
	at org.apache.accumulo.core.client.impl.ServerClient.execute(ServerClient.java:64)
	... 11 more
```
After reading the code, it looks like Gaffer will use PasswordToken for Accumulo Store and there is no other option, for example, KerberosToken. Do I miss anything here?  
```java
    /**
     * Creates a connection to an accumulo instance using the provided
     * parameters
     *
     * @param instanceName the instance name
     * @param zookeepers   the zoo keepers
     * @param userName     the user name
     * @param password     the password
     * @return A connection to an accumulo instance
     * @throws StoreException failure to create an accumulo connection
     */
    public static Connector getConnector(final String instanceName, final String zookeepers, final String userName,
                                         final String password) throws StoreException {
        final Instance instance = new ZooKeeperInstance(instanceName, zookeepers);
        try {
            return instance.getConnector(userName, new PasswordToken(password));
        } catch (final AccumuloException | AccumuloSecurityException e) {
            throw new StoreException("Failed to create accumulo connection", e);
        }
    }
```

**Comment by d47853** [comment]
Okay. I was wondering whether it would try to do something clever like convert the password token to a kerberos token. I guess it depends on how badly you want to use Kerberos with Accumulo + Gaffer. It looks like as is, we'd need a new feature to provide a Kerberos token in place of a password token for the Accumulo store. Feel free to add a PR if that's something you'd be interested in.

Alternatively it looks like you can disable SASL and use standard auth using the accumulo client.conf file.

```java
saslSupplier = new Supplier<SaslConnectionParams>() {
      @Override
      public SaslConnectionParams get() {
        // Use the clientConf if we have it
        if (null != clientConf) {
          if (!clientConf.hasSasl()) { // <-------- Disable this by setting "instance.rpc.sasl.enabled" to "false"
            return null;
          }
          return new SaslConnectionParams(clientConf, getCredentials().getToken());
        }
        AccumuloConfiguration conf = getConfiguration();
        if (!conf.getBoolean(Property.INSTANCE_RPC_SASL_ENABLED)) {
          return null;
        }
        return new SaslConnectionParams(conf, getCredentials().getToken());
      }
    };
```

**Comment by tdong6** [comment]
@d47853  Thanks, I disabled SASL and the example code just hangs. I can see it reached zookeeper from the zookeeper log and the connection keeps being started and stopped.
After I exited the code, I got 
```
EndOfStreamException: Unable to read additional data from client sessionid 0x17624cb458600f0, likely client has closed socket
        at org.apache.zookeeper.server.NIOServerCnxn.doIO(NIOServerCnxn.java:231)
        at org.apache.zookeeper.server.NIOServerCnxnFactory.run(NIOServerCnxnFactory.java:208)
        at java.lang.Thread.run(Thread.java:748)
```
which is expected, though I am not sure why it's stuck there. Then I realised the Gaffer I am using is built on top of Accumulo 1.9.3 and I am using a different version of Accumulo (1.7.2), so that could be the cause. At this point, I feel the best way for me to try Gaffer on Accumulo is probably setting up an Accumulo cluster (1.9.3) with default settings. 

I find there is a [discussion](https://github.com/gchq/Gaffer/issues/2201) on upgrading Gaffer to support Accumulo 2.x but don't see much follow-up if I am not missing anything. Is this still on the roadmap?
